stdlib/public/runtime/StaticBinaryELF.cpp

```diff
@@ -0,0 +1,329 @@
+        }
+
+        fullPathName = "/" + fname.str();
+        break;
+      }
+      if (line) {
+        free(line);
+      }
+      fclose(fp);
+    }
+  }
+
+
+  // Parse the ELF binary using mmap for the section headers, symbol table and
+  // string table.
+  void mmapExecutable() {
+    struct stat buf;
+    int fd = open(fullPathName.c_str(), O_RDONLY);
+    if (fd < 0) {
+      return;
+    }
+
+    if (fstat(fd, &buf) != 0) {
+      close(fd);
+      return;
+    }
+
+    // Map in the section headers.
+    size_t sectionHeadersSize = (elfHeader.e_shentsize * elfHeader.e_shnum);
+    sectionHeaders = new Mapping(fd, buf.st_size, elfHeader.e_shoff,
+                                 sectionHeadersSize);
+    if (sectionHeaders->mapping) {
+      auto section = findSectionHeader(SHT_SYMTAB);
+      if (section) {
+        assert(section->sh_entsize == sizeof(Elf_Sym));
+        symbolTableData = new Mapping(fd, buf.st_size, section->sh_offset,
+                                      section->sh_size);
+        if (symbolTableData->mapping) {
+          symbolTable = section;
+        }
+      }
+      section = findSectionHeader(SHT_STRTAB);
+      if (section) {
+        stringTableData = new Mapping(fd, buf.st_size, section->sh_offset,
+                                      section->sh_size);
+        if (stringTableData->mapping) {
+          stringTable = section;
+        }
+      }
+    }
+    close(fd);
+    return;
+  }
+
+  // Find the section header of a specified type in the section headers table.
+  const Elf_Shdr *findSectionHeader(Elf_Word sectionType) {
+    if (sectionHeaders && sectionHeaders->mapping) {
+      const ArrayRef<Elf_Shdr> headers = sectionHeaders->data<Elf_Shdr>();
+      for (size_t idx = 0; idx < elfHeader.e_shnum; idx++) {
+        if (idx == elfHeader.e_shstrndx) {
+          continue;
+        }
+        auto header = &headers[idx];
+        if (header->sh_type == sectionType) {
+          if (header->sh_entsize > 0 && header->sh_size % header->sh_entsize) {
+            fprintf(stderr,
+                    "section size is not a multiple of entrysize (%ld/%ld)\n",
+                    header->sh_size, header->sh_entsize);
+            return nullptr;
+          }
+          return header;
+        }
+      }
+    }
+    return nullptr;
+  }
+};
+
+
+static swift::Lazy<StaticBinaryELF> TheBinary;
+
+int
+swift::lookupSymbol(const void *address, SymbolInfo *info) {
+  // The pointers returned point into the mmap()'d binary so keep the
+  // object once instantiated.
+  auto &binary = TheBinary.get();
+
+  info->fileName = binary.getPathName();
+  info->baseAddress = binary.getSectionLoadAddress(address);
+
+  auto symbol = binary.findSymbol(address);
+  if (symbol != nullptr) {
+    info->symbolAddress = reinterpret_cast<void *>(symbol->st_value);
+    info->symbolName = binary.symbolName(symbol);
+  } else {
+    info->symbolAddress = nullptr;
+    info->symbolName = nullptr;
+  }
+
+  return 1;
```

Comment:
The stack trace dumper does not print a line for any lookup that fails so if all of them fail then no stack trace is printed at all. Returning an empty lookup structure and a result of 1 at least prints a stack even if the addresses aren't resolved and the depth of the stack can be seen.

I had a look at glibc's source and that always returns 1 even if lookup fails so I changed it to copy that behaviour . However I have no strong position on it and don't mind changing it back to zero